pkg/health/health.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	ctx "context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+)
+
+type cephStatus struct {
+	PgMap  pgMap        `json:"pgmap"`
+	Health healthStatus `json:"health"`
+}
+
+type healthStatus struct {
+	Status string `json:"status"`
+}
+
+type pgMap struct {
+	PgsByState []PgStateEntry `json:"pgs_by_state"`
+}
+
+type PgStateEntry struct {
+	StateName string `json:"state_name"`
+	Count     int    `json:"count"`
+}
+
+func Health(context *k8sutil.Context, operatorNamespace, clusterNamespace string) {
+
+	fmt.Println("Info: Checking if at least three mon pods are running on different nodes")
+	checkPodsOnNodes(context, clusterNamespace, "app=rook-ceph-mon")
+
+	fmt.Println()
+	fmt.Println("Info: Checking mon quorum and ceph health details")
+	checkMonQuorum(context, operatorNamespace, clusterNamespace)
+
+	fmt.Println()
+	fmt.Println("Info: Checking if at least three osd pods are running on different nodes")
+	checkPodsOnNodes(context, clusterNamespace, "app=rook-ceph-osd")
+
+	fmt.Println()
+	CheckAllPodsStatus(context, operatorNamespace, clusterNamespace)
+
+	fmt.Println("Info: Checking placement group status")
+	checkPgStatus(context, operatorNamespace, clusterNamespace)
+
+	fmt.Println()
+	fmt.Println("Info: Checking if at least one mgr pod is running")
+	checkMgrPodsStatusAndCounts(context, clusterNamespace)
+}
+
+func checkPodsOnNodes(context *k8sutil.Context, clusterNamespace, label string) {
+	var daemonType string
+	if strings.Contains(label, "osd") {
+		daemonType = "osd"
+	} else if strings.Contains(label, "mon") {
+		daemonType = "mon"
+	}
+
+	opts := metav1.ListOptions{LabelSelector: label}
+	podList, err := context.Clientset.CoreV1().Pods(clusterNamespace).List(ctx.TODO(), opts)
+	if err != nil {
+		fmt.Printf("\nfailed to list %s pods with label %s: %v\n", daemonType, opts.LabelSelector, err)
```

Comment:
return on error? The `podList` isn't valid.

Reply:
right, done